# DL4NLP SS17 Home Exercise 03
----------------------------------
**Due until Tuesday, 09.05. at 13:00**

## Task 1 Backpropagation (4P)

![MLP](backprop_hex03_task.png)

#### Network details

The activation function used in each layer is shown above the respective layer.

The network loss is square loss:
\begin{equation}
    \ell(t, y) = (t - y)^2 \qquad \ell'(t, y) = 2(y-t)
\end{equation}
$t$ denotes the true label, $y$ denotes the network output.

The weight matrices are initialized as follows:
\begin{equation}
	\mathbf{T} = \begin{pmatrix}
		0.19 & -0.92  \\
		-0.42 & -0.28 \\
	\end{pmatrix}
	\quad
	\mathbf{U} = \begin{pmatrix}
		0.61 \\
		-1.5 \\
	\end{pmatrix}
	\quad
	\mathbf{V} = \begin{pmatrix}
		1.5 & -0.81 & -0.24 \\
	\end{pmatrix}
	\quad
	\mathbf{W} = \begin{pmatrix}
		-1.4 & -0.81 \\
		-2.2 & -1.7 \\
		-0.27 & -0.73 \\
	\end{pmatrix}
\end{equation}

### Task 1.1 Backpropagation by hand (4P)
Using pen and paper, perform one execution of backpropagation on multilayer perceptron given above. "One execution" includes the forward propagation step, the backwards propagation and weight updates.
Use $\mathbf{x} = (-1, 1)$ as the input and $\mathbf{t} = (0,1)$ as the output. Use $\alpha=0.75$ for the learning rate.

The necessary formulas can be found on slides 11-13 [in the exercise slides](https://moodle.informatik.tu-darmstadt.de/mod/resource/view.php?id=7422). Round your results to four decimal points after each calculation.

#### Hints on the Submission Format
Please show your process of calculation. Both LaTeX solutions and scanned sheets are accepted.

## Task 2 Sentiment Polarity in Movie Reviews (6P)
We revisit last week's task of identifying sentiments in movie reviews.

#### Data (same as last week)
In the `hex03_data` archive, you can find a training, development and test dataset. Each line in these datasets has three entries, separated by a tab character (`\t`). The first is the movie review (available only for reference), the second is the sentiment label (`POS`itive or `NEG`ative). To facilitate the task, the third entry is a 100-dimensional vector representing the review (we'll cover in later lectures on *word embeddings* how this sentence representation has been generated).

#### Hints on the Submission Format
* Please submit your python code for all the tasks where it is applicable. Make sure to include comments explaining complicated/non-obvious sections of your code.
* Please also submit a copy of the console output of your code execution. Your code might run in 10 minutes on your watercooled battlestation, but it might not run in 10 minutes for the person who corrects your home exercises. Thank you!


### Task 2.0 Dataset reader (0P)
To read the datasets, reuse your own reader from last week or copy the one from the solution provided in moodle.

### Task 2.1 MLP in TensorFlow (2P)

Implement a multilayer perceptron in TensorFlow. Try to keep it as configurable as possible for the hyperparameter optimization for task 2.2.

Report loss and accuracy on the dev set for the following configuration:
* square loss
* 2 hidden layers, both with a dimension of 10 and using tanh activation
* weights initialized by the standard normal distribution (mean 0, sigma 1)
* batch size 10
* learning rate $\alpha = 0.01$
* 100 epochs


### Task 2.2 Hyperparameter Optimization (2P)
In lecture 02, slide 19ff you learned about random hyperparameter optimization. Try to find the best configuration for your MLP from task 2.1 by experimenting with the following hyperparameters:

* **A**: Known from the previous home exercise:
    * batch size
    * learning rate
    * number of training epochs


* **B**: New parameters:
    * loss function: square loss, hinge loss, softmax + cross-entropy loss
    * number of hidden layers: anything values from the interval $[0, \infty[$
    * hidden layer dimension: anything values from the interval $[1, \infty[$
    * activation function(s): sigmoid, tanh, ReLU, softplus, leaky ReLU, ELU, maxout
    * 2-norm regularization of hidden layer weights (see lecture 03, slide 37): yes/no
    * optimizer: SGD, Adam, AdaGrad, ...

To obtain the 2P for this task, you need to experiment with all three parameters from set **A** and at least three parameters from set **B**. Report your top 10 parameter sets and their loss/accuracy on the dev set. Then, report loss and accuracy of your best parameter set on the test set.

Hints:
* Activation functions and hidden layer dimensions can also be chosen differently for each layer.
* When varying the loss function or the activation functions, it might make sense to redefine the truth labels in the datasets from 0 for NEGative and 1 for POSitive to something else (think of the co-domain of activation functions).

### Task 2.3 MLP in Keras (2P)
[Install Keras](https://keras.io/#installation) on your machine and implement an MLP with Keras. Hardcode the initial hyperparameter configuration mentioned in Task 2.1 and report loss and accuracy on the dev set.

# Answer
## Task 1 Backpropagation (4P)

Forward propagation:

$\mathbf{r} = max(0,\mathbf{x} \mathbf{T}) =max(0, (-1, 1) \begin{pmatrix}
		0.19 & -0.92  \\
		-0.42 & -0.28 \\
	\end{pmatrix}) = (0.61, 0.64) $ 
    
$\mathbf{q} = tanh(\mathbf{r} \mathbf{U}) = tanh((0.61, 0.64)\begin{pmatrix}
		0.61 \\
		-1.5 \\
	\end{pmatrix}) = tanh(-0.5879) = -0.67$

$\mathbf{p}=sig(\mathbf{q} \mathbf{V})=sig(-0.67\begin{pmatrix}
		1.5 & -0.81 & -0.24 \\
	\end{pmatrix})=sig(\begin{pmatrix}-1.005 & 0.5427 & 0.1608\\
\end{pmatrix}) =\begin{pmatrix} 0.27 &  0.63 & 0.54 \end{pmatrix}  $
    
$\mathbf{y}=sig(\mathbf{p}\mathbf{W}=\begin{pmatrix} 0.27 &  0.63 & 0.54 \end{pmatrix}
\begin{pmatrix}
		-1.4 & -0.81 \\
		-2.2 & -1.7 \\
		-0.27 & -0.73 \\
	\end{pmatrix})=sig(\begin{pmatrix}-2.22 & -1.68\end{pmatrix})=\begin{pmatrix} 0.098 & 0.16\end{pmatrix}$
    
$\mathbf{E}=(\mathbf{t} - \mathbf{y})^2=\begin{pmatrix} 0.204 & 1.198 \end{pmatrix} $

$\frac{d \mathbf{E}}{dy} =2(\mathbf{t} - \mathbf{y}) =\begin{pmatrix} 0.408 &  2.396 \end{pmatrix}$
    


## Task 2.1 & 2.2 MLP in Tensorflow & Hyperparameter Optimization

In [2]:
import numpy as np
import os
import tensorflow as tf
import progressbar as pb

label_pos = 1
label_neg = 0

data_dim = 100
data_dim_with_bias = data_dim + 1

# learning_rate = 0.1
learning_rate = 0.01
# learning_rate = 0.001
batch_size = 5
# batch_size = 10
# batch_size = 20
# epochs = 50
epochs = 100
# epochs = 150

def read_dataset(src):
    """Reads a dataset from the specified path and returns input vectors and labels in an array of shape (n, 101)."""
    with open(src, 'r') as src_file:
        # preallocate memory for the data
        num_lines = sum(1 for line in src_file)
        data = np.empty((num_lines, data_dim_with_bias), dtype=np.float16)
        labels = np.empty((num_lines, 1), dtype=np.float16)

        # reset the file pointer to the beginning of the file
        src_file.seek(0)
        for i, line in enumerate(src_file):
            _, str_label, str_vec = line.split('\t')
            labels[i] = label_pos if str_label.split('=')[1] == "POS" else label_neg
            data[i,:data_dim] = [float(f) for f in str_vec.split()]
            data[i,data_dim] = 1
    return data, labels

def get_dataset(src_folder, name="train"):
    path = os.path.join(src_folder, "rt-polarity.{}.vecs".format(name))
    return read_dataset(path)

def get_random_batches(X, y, batch_size):
    perm = np.random.permutation(len(y))
    X = X[perm]
    y = y[perm]
    # when using array_split for 100 datapoints and batch size 33 one would get batches [33, 33, 33, 1]
    X_batches = np.array_split(X, len(y)//batch_size)
    y_batches = np.array_split(y, len(y)//batch_size)
    return X_batches, y_batches

# load the data
train_x, train_y, = get_dataset("DATA", "train")
dev_x, dev_y = get_dataset("DATA", "dev")
test_x, test_y = get_dataset("DATA", "test")

In [6]:
import tensorflow as tf
n_input = data_dim_with_bias
n_hidden_1 = 10
n_hidden_2 = 8

# define weights
def init_weights(n_input,n_hidden_1,n_hidden_2):
    weights = {
        'h1': tf.Variable(tf.random_normal(shape=[n_input,n_hidden_1],mean=0,stddev=1)),
        'h2': tf.Variable(tf.random_normal(shape=[n_hidden_1,n_hidden_2],mean=0,stddev=1)),
        'out': tf.Variable(tf.random_normal(shape=[n_hidden_2,1],mean=0,stddev=1))}
    return weights

# define model
def MLP(X, weights):
    # first hidden layer
    layer_1 = tf.tanh(tf.matmul(X,weights['h1']))
#     layer_1 = tf.nn.relu(tf.matmul(X,weights['h1']))
    # second hidden layer
    layer_2 = tf.tanh(tf.matmul(layer_1,weights['h2']))
#     layer_2 = tf.nn.relu(tf.matmul(layer_1,weights['h2']))
    # output
    out = tf.nn.softmax(tf.matmul(layer_2,weights['out']))
#     out = tf.sigmoid(tf.matmul(layer_2, weights['out']))
    return out

# define tf graph
X = tf.placeholder(tf.float32, shape=(None, n_input))
y = tf.placeholder(tf.float32, shape=(None, 1))

# initialize weights
weights = init_weights(n_input,n_hidden_1,n_hidden_2)

# calculate prediction
pred = MLP(X,weights)

# calculate cost function and optimization 
# cost = tf.losses.mean_squared_error(y,pred)
cost = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(logits=pred, labels=y))

# train_op = tf.train.GradientDescentOptimizer(learning_rate).minimize(cost)
train_op = tf.train.AdamOptimizer(learning_rate=learning_rate).minimize(cost)

tf.set_random_seed(42)

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())    
    for m in range(epochs):
        train_x_batches, train_y_batches = get_random_batches(train_x, train_y, batch_size)
        for X_batch, y_batch in zip(train_x_batches, train_y_batches):
            sess.run(train_op,feed_dict={X:X_batch,y:y_batch})
    #print results on dev and test
    correct_prediction = tf.equal(tf.round(pred),y)
    accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
    loss = tf.losses.mean_squared_error(y,pred)
   
    print("Accuracy on test-set: {0:.1%}".format(accuracy.eval({X:test_x,y:test_y})))
    print("Loss on test-set:",loss.eval({X:test_x,y:test_y}))
    print("Accuracy on dev-set: {0:.1%}".format(accuracy.eval({X:dev_x,y:dev_y})))
    print("Loss on dev-set:",loss.eval({X:dev_x,y:dev_y}))

Accuracy on test-set: 49.7%
Loss on test-set: 0.50344
Accuracy on dev-set: 50.9%
Loss on dev-set: 0.490932


## Top 10 parameter sets:
| TOP 10 | batch size | learn rate | epochs | Loss function | # of hidden layers | Activation fucntion | Relularization | Optimizer | Accuracy | Loss        
| :- |-------------: | :-:
| 1 | 5 | 0.01 | 100 | square loss | 10, 8 | tanh | No |SGD |  68.5% | 0.217952
| 2 | 5 | 0.01 | 100 | square loss | 10, 10 | tanh | No |SGD | 66.2% | 0.220211
| 3 | 5 | 0.01 | 100 | square loss | 10, 5 | tanh | No |SGD | 66.2% | 0.226095
| 4 | 5 | 0.01 | 200 | square loss | 10, 10 | tanh | No |SGD | 65.7% | 0.213632
| 5 | 5 | 0.01 | 100 | square loss | 10, 8 | tanh & softplus | No | SGD | 65.3% | 0.215742
| 6 | 10 | 0.01 | 100 | square loss | 10, 10 | tanh | No |SGD | 62.7% | 0.228515
| 7 | 10 | 0.1 | 100 | square loss | 10, 10 | tanh | No |SGD | 61.5% | 0.232456
| 8 | 20 | 0.01 | 100 | square loss | 10, 10 | tanh | No |SGD | 59.2% | 0.251632
| 9 | 5 | 0.01 | 50 | square loss | 10, 10 | tanh | No |SGD | 58.3% | 0.247174
| 10 | 5 | 0.01 | 100 | cross entropy loss | 10, 8 | softmax | No |SGD | 50.9% | 0.490932


## Best parameter on test-sets:
Accuracy on test-set: 67.3% --- Loss on test-set: 0.21821

### All Results
Accuracy on test-set: 63.0% --- Loss on test-set: 0.2239

Accuracy on dev-set: 62.7% --- Loss on dev-set: 0.228515

###  When learning_rate= 0.001
Accuracy on test-set: 55.3% --- Loss on test-set: 0.250549

Accuracy on dev-set: 52.3% --- Loss on dev-set: 0.258009

###  When learning_rate= 0.1  
Accuracy on test-set: 62.9% --- Loss on test-set: 0.226974

Accuracy on dev-set: 61.5% --- Loss on dev-set: 0.232456

### When batch_size= 5
Accuracy on test-set: 67.4% --- Loss on test-set: 0.208212

Accuracy on dev-set: 66.2% --- Loss on dev-set: 0.220211


### When batch_size= 20
Accuracy on test-set: 59.7% --- Loss on test-set: 0.2413

Accuracy on dev-set: 59.2% --- Loss on dev-set: 0.251632

### When epochs= 50
Accuracy on test-set: 56.7% --- Loss on test-set: 0.25013

Accuracy on dev-set: 58.3% --- Loss on dev-set: 0.247174

### When epochs=200
Accuracy on test-set: 65.5% --- Loss on test-set: 0.214483

Accuracy on dev-set: 65.7% --- Loss on dev-set: 0.213632

### When hidden layer dimension: 10, 5
Accuracy on test-set: 67.0% --- Loss on test-set: 0.22078

Accuracy on dev-set: 66.2% --- Loss on dev-set: 0.226095

### When hidden layer diemension: 10, 8
Accuracy on test-set: 67.3% --- Loss on test-set: 0.21821

Accuracy on dev-set: 68.5% --- Loss on dev-set: 0.217952

### Activation: ReLU, ReLU, tanh
Accuracy on test-set: 49.5% --- Loss on test-set: 0.509068

Accuracy on dev-set: 50.8% --- Loss on dev-set: 0.49531

### Activation: tanh, tanh, softplus
Accuracy on test-set: 65.5% --- Loss on test-set: 0.220412

Accuracy on dev-set: 65.3% --- Loss on dev-set: 0.215742

### Loss function: softmax+cross-entropy
Accuracy on test-set: 49.7% --- Loss on test-set: 0.50344

Accuracy on dev-set: 50.9% --- Loss on dev-set: 0.490932

## Task 2.3 MLP in Keras

In [9]:
from keras.models import Sequential
from keras.layers import Dense, Activation

n_hidden_1 = 10
n_hidden_2 = 8
batch_size = 10
epochs = 100

model = Sequential()

model.add(Dense(n_hidden_1,activation='tanh', input_dim=data_dim_with_bias))
# model.add(Activation('tanh'))

model.add(Dense(n_hidden_2,activation='tanh'))
# model.add(Activation('tanh'))

model.add(Dense(1,activation='sigmoid'))
# model.add(Activation('sigmoid'))


model.compile(loss='binary_crossentropy',
             optimizer='sgd',
             metrics=['accuracy'])

model.fit(train_x, train_y,epochs=epochs,batch_size=batch_size)
scores = model.evaluate(dev_x,dev_y,batch_size=batch_size)
print("Accuracy on dev-set: %.2f%%" % (scores[1]*100))  # accuracy
print("Loss on dev-set:",scores[0]) # Loss


Epoch 1/100
7464/7464 [==============================] - 0s - loss: 0.6870 - acc: 0.5588     
Epoch 2/100
7464/7464 [==============================] - 0s - loss: 0.6537 - acc: 0.6085     
Epoch 3/100
7464/7464 [==============================] - 0s - loss: 0.6324 - acc: 0.6400     
Epoch 4/100
7464/7464 [==============================] - 0s - loss: 0.6203 - acc: 0.6526     
Epoch 5/100
7464/7464 [==============================] - 0s - loss: 0.6061 - acc: 0.6652     
Epoch 6/100
7464/7464 [==============================] - 0s - loss: 0.6035 - acc: 0.6736     
Epoch 7/100
7464/7464 [==============================] - 0s - loss: 0.6028 - acc: 0.6677     
Epoch 8/100
7464/7464 [==============================] - 0s - loss: 0.6026 - acc: 0.6748     
Epoch 9/100
7464/7464 [==============================] - 0s - loss: 0.6011 - acc: 0.6760     
Epoch 10/100
7464/7464 [==============================] - 0s - loss: 0.5936 - acc: 0.6787     
Epoch 11/100
7464/7464 [==============================] - 0

7464/7464 [==============================] - 0s - loss: 0.5457 - acc: 0.7142     
Epoch 88/100
7464/7464 [==============================] - 0s - loss: 0.5401 - acc: 0.7244     
Epoch 89/100
7464/7464 [==============================] - 0s - loss: 0.5505 - acc: 0.7121     
Epoch 90/100
7464/7464 [==============================] - 0s - loss: 0.5445 - acc: 0.7207     
Epoch 91/100
7464/7464 [==============================] - 0s - loss: 0.5460 - acc: 0.7231     
Epoch 92/100
7464/7464 [==============================] - 0s - loss: 0.5474 - acc: 0.7169     
Epoch 93/100
7464/7464 [==============================] - 0s - loss: 0.5446 - acc: 0.7168     
Epoch 94/100
7464/7464 [==============================] - 0s - loss: 0.5656 - acc: 0.7004     
Epoch 95/100
7464/7464 [==============================] - 0s - loss: 0.5602 - acc: 0.6858     
Epoch 96/100
7464/7464 [==============================] - 0s - loss: 0.5584 - acc: 0.7098     
Epoch 97/100
7464/7464 [==============================] - 0s - 

As we can see above, Accuracy on dev-set: 70.04%, Loss on dev-set: 0.571630382515

In [7]:
import keras
print(keras.__version__)

2.0.4
